ddtrace/tracer/tracer.go

```diff
@@ -323,27 +323,40 @@ func (t *tracer) worker(tick <-chan time.Time) {
 // finishedTrace holds information about a trace that has finished, including its spans.
 type finishedTrace struct {
 	spans    []*span
-	decision samplingDecision
+	willSend bool // willSend indicates whether the trace will be sent to the agent.
 }
 
 // sampleFinishedTrace applies single-span sampling to the provided trace, which is considered to be finished.
 func (t *tracer) sampleFinishedTrace(info *finishedTrace) {
-	if info.decision == decisionKeep {
+	// The trace is kept, no need to run single span sampling rules.
+	// It is important that to check the sampling decision before checking the
+	// availability of single span sampling rules.
+	if len(info.spans) > 0 {
+		if p, ok := info.spans[0].context.samplingPriority(); ok && p > 0 {
```

Comment:
Why do we look at the samplingPriority here instead of looking at the `info.willSend` value?

Reply:
Because the sampling priority tells us whether the trace is kept(if it is, no need to run single spand sampling). But the willSend only tells us whether this trace, regardless of the sampling decision, is going to be sent to the agent.

Reply:
Sorry I'm also confused. When will `willSend` be false if the sampling priority is indicating keep? I thought we were supposed to only use the sampling decision once it's calculated, and not pay attention to the sampling priority anymore. It seems more like a bug if we can't rely on the sampling decision here to know whether or not something is about to be sent to the agent.